File: IT/1/Stringhe/Esercizi/Java/Esercizio30.java
/* file Esercizio30.java */

import java.io.Console;
import java.util.Scanner;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class Esercizio30 {
	public static void main(String [] args) {
		System.out.println("Programmiamo un client di posta elettronica.");

		// destinatario
		String dest = "";

		// oggetto
		String obj = "";

		// corpo della mail
		String body = "";

		// eventuale allegato
		String att = "";

		boolean exit = false;

		System.out.println("From: joebloggs@programmiamo.com");
		Console input = System.console();

		String r = "[\\w\\.]+@[\\w]+\\.[\\w]{2,}";
		Pattern mailPattern = Pattern.compile(r);

		Pattern attPattern = Pattern.compile("alleg(at)?o", Pattern.CASE_INSENSITIVE);

		while (!exit) {
			if (dest.isEmpty()) {
				String _dest = input.readLine("To: ");
				if (!mailPattern.matcher(_dest).matches()) continue;
				dest = _dest;
			}

			if (obj.isEmpty()) {
				String _obj = input.readLine("Oggetto: ");
				obj = !_obj.isEmpty() ? _obj : " ";
				/*
				if (!obj.isEmpty()) obj = _obj;
				else {
					String placeHolder = "<nessun oggetto>";
					System.out.print(placeHolder);
				}
				*/
			}

			if (body.isEmpty()) {
				System.out.println("------- Messaggio -------" + System.lineSeparator());
				// String _body = input.readLine();
				Scanner sc = new Scanner(input.reader());
				StringBuilder sb = new StringBuilder();

				// si chiude con CTRL+D
				while (sc.hasNextLine()) {
					sb.append(sc.nextLine());
				}

				String _body = sb.toString();
				body = !_body.isEmpty() ? _body : " ";
				System.out.println(System.lineSeparator() + "-------------------------");
			}

			System.out.println("\"y\" - invia la mail" + System.lineSeparator() + "\"a\" - aggiungi un allegato" + System.lineSeparator() + "\"q\" - esci");
			String buf1 = input.readLine();
			char opt = !buf1.isEmpty() ? Character.toLowerCase(buf1.charAt(0)) : Character.MIN_VALUE;

			if (opt == 'y') {
				if (att.isEmpty()) {
					Matcher matcher1 = attPattern.matcher(obj);
					Matcher matcher2 = attPattern.matcher(body);

					if (matcher1.find() || matcher2.find()) {
						System.out.println("Forse vuoi allegare qualcosa. Inviare?" + System.lineSeparator() + "\"s\" - invia senza allegati" + System.lineSeparator() + "\"n\" - torna alla mail");
						String buf2 = input.readLine();
						char c = !buf2.isEmpty() ? Character.toLowerCase(buf2.charAt(0)) : 'n';
						if (c != 's') continue;
					}
				}

				System.out.println("Messaggio inviato.");
			}
			else if (opt == 'a') {
				String _att = input.readLine("File da allegare: ");
				if (_att.isEmpty()) continue;
				att = _att;
				System.out.println(System.lineSeparator() + "Caricamento di \"" + att + "\"...");
				System.out.println("####### 100% completato #######" + System.lineSeparator());
				continue;
			}

			exit = true;
		}
	}
}

File: IT/1/Stringhe/Esercizi/C++/esercizio30.cpp
/*
file esercizio30.cpp
compilare con: g++ -std=c++11 esercizio30.cpp
*/

#include <iostream>
#include <regex>

using std::cin;
using std::string;
using std::cout;
using std::endl;
using std::regex;
using std::numeric_limits;
using std::regex_constants::icase;

int main()
{
	cout << "Programmiamo un client di posta elettronica." << endl;

	// destinatario
	string dest = "";

	// oggetto
	string obj = "";

	// corpo della mail
	string body = "";

	// eventuale allegato
	string att = "";

	bool exit = false;

	cout << "From: joebloggs@programmiamo.com" << endl;

	regex mailPattern("[\w\.]+@[\w]+\.[\w]{2,}");
	regex attPattern("alleg(at)?o", icase);

	while (!exit) {
		if (dest.empty()) {
			cout << "To: ";
			string _dest;
			getline(cin, _dest);
			if (!regex_match(_dest, mailPattern)) continue;
			dest = _dest;
		}

		if (obj.empty()) {
			cout << "Oggetto: ";
			string _obj;
			getline(cin, _obj);
			obj = !_obj.empty() ? _obj : " ";
		}

		if (body.empty()) {
			cout << "------- Messaggio -------" << endl << endl;
			string _body;
			string line;

			// si chiude con CTRL+D
			while (!getline(cin, line).eof()) {
				_body += line + "\n";
			}

			cin.clear();

			body = !_body.empty() ? _body : " ";
			cout << endl << "-------------------------" << endl;
		}

		cout << "\"y\" - invia la mail" << endl << "\"a\" - aggiungi un allegato" << endl << "\"q\" - esci" << endl;
		string buf1;
		getline(cin, buf1);
		if (buf1.size() > 1) continue;
		char opt = !buf1.empty() ? tolower(buf1[0]) : numeric_limits<char>::min();

		if (opt == 'y') {
			if (att.empty()) {
				if (regex_search(obj, attPattern) || regex_search(body, attPattern)) {
					cout << "Forse vuoi allegare qualcosa. Inviare?" << endl << "\"s\" - invia senza allegati" << endl << "\"n\" - torna alla mail" << endl;
					string buf2;
					getline(cin, buf2);
					char c = !buf2.empty() ? tolower(buf2[0]) : numeric_limits<char>::min();
					if (c != 's') continue;
				}
			}

			cout << "Messaggio inviato." << endl;
		}
		else if (opt == 'a') {
			cout << "File da allegare: ";
			string _att;
			getline(cin, _att);

			if (_att.empty()) continue;
			att = _att;
			cout << endl << "Caricamento di \"" << att << "\"..." << endl;
			cout << "####### 100% completato #######" << endl;
			continue;
		}

		exit = true;
	}
}

